generators/E906LegacyVtxGen/SQPrimaryVertexGen.C

```diff
@@ -0,0 +1,270 @@
+  beamProfile->SetNpy(200); 
+ 
+  beamProfile->SetParameter(0, rc->get_DoubleFlag("X_BEAM"));
+  beamProfile->SetParameter(1, rc->get_DoubleFlag("SIGX_BEAM"));
+  beamProfile->SetParameter(2, rc->get_DoubleFlag("Y_BEAM"));
+  beamProfile->SetParameter(3, rc->get_DoubleFlag("SIGY_BEAM"));
+
+  rndm.SetSeed(PHRandomSeed());
+
+  //If inited from a root file, then skip this step, so that it can be used independent of the Fun4All
+  if(geoManager == nullptr) geoManager = PHGeomUtility::GetTGeoManager(topNode);
+
+  //Read the default 
+  defaultMatProf = new MaterialProfile;
+  fillMaterialProfile(defaultMatProf, 0., 0.);
+
+  inited = true;
+}
+
+void SQPrimaryVertexGen::fillMaterialProfile(MaterialProfile* prof, double xvtx, double yvtx)
+{
+  //std::cout << xvtx << "  " << yvtx << std::endl;
+  //retrieve the z position of all geometry boundaries
+  recoConsts* rc = recoConsts::instance();
+
+  double z[100];
+  int nBoundaries = 0;
+
+  double eps = 1.E-6;
+  double z_curr = -800.;//not beyond collimator(z= -602.36 cm,lenth=121.92 cm) @Abi //-500.;
+  z[0] = z_curr;
+  while(z_curr < 503.)
+    {
+      geoManager->IsSameLocation(xvtx, yvtx, z_curr, true);
+      geoManager->SetCurrentDirection(0., 0., 1.);
+
+      TGeoNode* node = geoManager->FindNextBoundary(600.);
+      double z_step  = geoManager->GetStep();
+
+      if(z_step > eps)
+	{
+	  z_curr += z_step;
+	  z[++nBoundaries] = z_curr; 
+	}
+      else
+	{
+	  z_curr += eps;
+	}
+    }
+
+  //use the z boundaries to create interactibles
+  for(int i = 0; i < nBoundaries; ++i)
+    {
+      double z_center = 0.5*(z[i] + z[i+1]);
+      if(z_center > 503.) continue;
+ 
+      SQBeamlineObject obj(geoManager->FindNode(xvtx, yvtx, z_center)->GetVolume()->GetMaterial());
+      TString name = geoManager->FindNode(xvtx, yvtx, z_center)->GetVolume()->GetName();
+      obj.z_up = z[i];
+      obj.z_down = z[i+1];
+      obj.z0 = z_center;
+      obj.length = z[i+1] - z[i];
+    
+      //TString name = obj.name;    
+      //std::cout<<" Name of beam volume "<<name<<std::endl;  
+
+      if (_targetOnlyMode){
+	if(! name.Contains("Target")) continue;
+      }
+
+      if (_dumpOnlyMode){
+	if(! name.Contains("fmag")) continue;
+      }
+
+      prof->interactables.push_back(obj);
+    }
+
+  //calculate all related properties
+  prof->calcProb();
+
+  // for(int i = 0; i < prof->nPieces; ++i)
+  // {
+  //   std::cout << prof->interactables[i] << std::endl;
+  // }
+  // return 0;
+}
+
+
+TVector3 SQPrimaryVertexGen::generateVertex()
+{
+  if(!inited) init();
+
+  double xvtx, yvtx, zvtx;
+  generateVtxPerp(xvtx, yvtx);
+
+  bool inTarget = ((xvtx-targetX0)*(xvtx-targetX0)/targetSX/targetSX + (yvtx-targetY0)*(yvtx-targetX0)/targetSY/targetSY) < 1.;
+  MaterialProfile* activeProf;
+  if(!inTarget)
+    {
+      
+      activeProf = new MaterialProfile;
+      fillMaterialProfile(activeProf, xvtx, yvtx);
+    }
+  else
+    {
+      activeProf = defaultMatProf;
```

Comment:
A real issue here is that `nPieces` is not updated in this "else" case, since `fillMaterialProfile` is not called.

I think `nPieces` should not be a member variable and instead `interactables.size()` should be called.

Reply:
Thank you Kenichi for careful look. Here, for the else case, the "nPieces" should be constant as the fillMaterialProfile() is called initially for the defaultMatProf in init() function.

Reply:
The issue should happen when the "if" case is executed in an event and later the "else" case is executed in another event.  `nPieces` is changed in the "if" case (for the outside of the target area), and then it is not changed back in the "else" case (for the inside).

Reply:
I see. Yes, you are right. I will look into it.